File: PdfClown-NetStandard/CliSamples/org/pdfclown/samples/cli/TextInfoExtractionSample.cs
using org.pdfclown.documents;
using org.pdfclown.documents.contents;
using org.pdfclown.documents.contents.composition;
using org.pdfclown.documents.contents.colorSpaces;
using org.pdfclown.documents.contents.objects;
using org.pdfclown.files;
using org.pdfclown.tools;

using System;
using System.Drawing;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to retrieve text content along with its graphic attributes
    (font, font size, text color, text rendering mode, text bounding box...) from a PDF document;
    it also generates a document version decorated by text bounding boxes.</summary>
  */
  public class TextInfoExtractionSample
    : Sample
  {
    private DeviceRGBColor[] textCharBoxColors = new DeviceRGBColor[]
      {
        new DeviceRGBColor(200 / 255d, 100 / 255d, 100 / 255d),
        new DeviceRGBColor(100 / 255d, 200 / 255d, 100 / 255d),
        new DeviceRGBColor(100 / 255d, 100 / 255d, 200 / 255d)
      };
    private DeviceRGBColor textStringBoxColor = DeviceRGBColor.Black;

    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(var file = new File(filePath))
      {
        Document document = file.Document;
  
        PageStamper stamper = new PageStamper(); // NOTE: Page stamper is used to draw contents on existing pages.
  
        // 2. Iterating through the document pages...
        foreach(Page page in document.Pages)
        {
          Console.WriteLine("\nScanning page " + page.Number + "...\n");
  
          stamper.Page = page;
  
          Extract(
            new ContentScanner(page), // Wraps the page contents into a scanner.
            stamper.Foreground
            );
  
          stamper.Flush();
        }
  
        // 3. Decorated version serialization.
        Serialize(file);
      }
    }

    /**
      <summary>Scans a content level looking for text.</summary>
    */
    /*
      NOTE: Page contents are represented by a sequence of content objects,
      possibly nested into multiple levels.
    */
    private void Extract(
      ContentScanner level,
      PrimitiveComposer composer
      )
    {
      if(level == null)
        return;

      while(level.MoveNext())
      {
        ContentObject content = level.Current;
        if(content is Text)
        {
          ContentScanner.TextWrapper text = (ContentScanner.TextWrapper)level.CurrentWrapper;
          int colorIndex = 0;
          foreach(ContentScanner.TextStringWrapper textString in text.TextStrings)
          {
            RectangleF textStringBox = textString.Box.Value;
            Console.WriteLine(
              "Text ["
                + "x:" + Math.Round(textStringBox.X) + ","
                + "y:" + Math.Round(textStringBox.Y) + ","
                + "w:" + Math.Round(textStringBox.Width) + ","
                + "h:" + Math.Round(textStringBox.Height)
                + "] [font size:" + Math.Round(textString.Style.FontSize) + "]: " + textString.Text
                );

            // Drawing text character bounding boxes...
            colorIndex = (colorIndex + 1) % textCharBoxColors.Length;
            composer.SetStrokeColor(textCharBoxColors[colorIndex]);
            foreach(TextChar textChar in textString.TextChars)
            {
              /*
                NOTE: You can get further text information
                (font, font size, text color, text rendering mode)
                through textChar.style.
              */
              composer.DrawRectangle(textChar.Box);
              composer.Stroke();
            }

            // Drawing text string bounding box...
            composer.BeginLocalState();
            composer.SetLineDash(new LineDash(new double[]{5}));
            composer.SetStrokeColor(textStringBoxColor);
            composer.DrawRectangle(textString.Box.Value);
            composer.Stroke();
            composer.End();
          }
        }
        else if(content is XObject)
        {
          // Scan the external level!
          Extract(
            ((XObject)content).GetScanner(level),
            composer
            );
        }
        else if(content is ContainerObject)
        {
          // Scan the inner level!
          Extract(
            level.ChildLevel,
            composer
            );
        }
      }
    }
  }
}

File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/TextInfoExtractionSample.java
package org.pdfclown.samples.cli;

import java.awt.geom.Rectangle2D;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.contents.ContentScanner;
import org.pdfclown.documents.contents.LineDash;
import org.pdfclown.documents.contents.TextChar;
import org.pdfclown.documents.contents.colorSpaces.DeviceRGBColor;
import org.pdfclown.documents.contents.composition.PrimitiveComposer;
import org.pdfclown.documents.contents.objects.ContainerObject;
import org.pdfclown.documents.contents.objects.ContentObject;
import org.pdfclown.documents.contents.objects.Text;
import org.pdfclown.documents.contents.objects.XObject;
import org.pdfclown.files.File;
import org.pdfclown.tools.PageStamper;

/**
  This sample demonstrates <b>how to retrieve text content along with its graphic attributes</b>
  (font, font size, text color, text rendering mode, text bounding box...) from a PDF document;
  it also generates a document version decorated by text bounding boxes.

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.8
  @version 0.1.2.1, 04/30/15
*/
public class TextInfoExtractionSample
  extends Sample
{
  private final DeviceRGBColor[] textCharBoxColors = new DeviceRGBColor[]
    {
      new DeviceRGBColor(200 / 255d, 100 / 255d, 100 / 255d),
      new DeviceRGBColor(100 / 255d, 200 / 255d, 100 / 255d),
      new DeviceRGBColor(100 / 255d, 100 / 255d, 200 / 255d)
    };
  private final DeviceRGBColor textStringBoxColor = DeviceRGBColor.Black;

  @Override
  public void run(
    )
  {
    // 1. Opening the PDF file...
    File file;
    {
      String filePath = promptFileChoice("Please select a PDF file");
      try
      {file = new File(filePath);}
      catch(Exception e)
      {throw new RuntimeException(filePath + " file access error.",e);}
    }
    Document document = file.getDocument();

    PageStamper stamper = new PageStamper(); // NOTE: Page stamper is used to draw contents on existing pages.

    // 2. Iterating through the document pages...
    for(Page page : document.getPages())
    {
      System.out.println("\nScanning page " + page.getNumber() + "...\n");

      stamper.setPage(page);

      extract(
        new ContentScanner(page), // Wraps the page contents into a scanner.
        stamper.getForeground()
        );

      stamper.flush();
    }

    // 3. Decorated version serialization.
    serialize(file);
  }

  /**
    Scans a content level looking for text.
  */
  /*
    NOTE: Page contents are represented by a sequence of content objects,
    possibly nested into multiple levels.
  */
  private void extract(
    ContentScanner level,
    PrimitiveComposer composer
    )
  {
    if(level == null)
      return;

    while(level.moveNext())
    {
      ContentObject content = level.getCurrent();
      if(content instanceof Text)
      {
        ContentScanner.TextWrapper text = (ContentScanner.TextWrapper)level.getCurrentWrapper();
        int colorIndex = 0;
        for(ContentScanner.TextStringWrapper textString : text.getTextStrings())
        {
          Rectangle2D textStringBox = textString.getBox();
          System.out.println(
            "Text ["
              + "x:" + Math.round(textStringBox.getX()) + ","
              + "y:" + Math.round(textStringBox.getY()) + ","
              + "w:" + Math.round(textStringBox.getWidth()) + ","
              + "h:" + Math.round(textStringBox.getHeight())
              + "] [font size:" + Math.round(textString.getStyle().getFontSize()) + "]: " + textString.getText()
            );

          // Drawing text character bounding boxes...
          colorIndex = (colorIndex + 1) % textCharBoxColors.length;
          composer.setStrokeColor(textCharBoxColors[colorIndex]);
          for(TextChar textChar : textString.getTextChars())
          {
            /*
              NOTE: You can get further text information
              (font, font size, text color, text rendering mode)
              through textChar.style.
            */
            composer.drawRectangle(textChar.getBox());
            composer.stroke();
          }

          // Drawing text string bounding box...
          composer.beginLocalState();
          composer.setLineDash(new LineDash(new double[]{5}));
          composer.setStrokeColor(textStringBoxColor);
          composer.drawRectangle(textString.getBox());
          composer.stroke();
          composer.end();
        }
      }
      else if(content instanceof XObject)
      {
        // Scan the external level!
        extract(
          ((XObject)content).getScanner(level),
          composer
          );
      }
      else if(content instanceof ContainerObject)
      {
        // Scan the inner level!
        extract(
          level.getChildLevel(),
          composer
          );
      }
    }
  }
}
